HPCC-17506 SPRAY VARIABLE without at least one TERMINATOR fails

AttilaVamos opened the pull request:
Add code to handle the "there is only one row/record and it is
terminated by EOF instead of TERMINATOR and the record structure
definition generation doesn't finish yet" case in the
CCsvPartitioner::getSplitRecordSize() method.

Create Regression Suite test case to test spray a file contains only one
record/row with and without terminator at the end situation.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually together with all other Spray and Despray test cases w/wo --pq.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 61ba9fb17aa76d59897e9e630d6954b816ae6145
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.456 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |724 |724 |0 |
|test (thor) |617 |617 |0 |
|test (roxie) |746 |746 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@jakesmith please review.

jakesmith commented:
@AttilaVamos - Looks good, a couple of minor comments.

richardkchapman commented:
Where are the comments?

jakesmith commented:
Whoops, didn't hit finish review button.

@AttilaVamos - see comments.

richardkchapman commented on dali/ft/daftformat.cpp:
```diff
@@ -928,6 +928,21 @@ size32_t CCsvPartitioner::getSplitRecordSize(const byte * start, unsigned maxToR
+        // by EOF instead of TERMINATOR therefore the record structure definition
+        // generation doesn't finish yet.
+        isFirstRow = false;
+        LOG(MCdebugProgressDetail, unknownJob, "%d (0x%08x) bytes processed, but there is no terminator found. It seems there is only one row/record and it is terminated by EOF instead of TERMINATOR.", numOfProcessedBytes, numOfProcessedBytes);
```
I think it's valid, though maybe unusual, to have just the heading. Not sure it's worth logging.

AttilaVamos replied:
Based on the RFC 4180 1.2.2 it is valid CSV file. I put thsi logging only to record it is happened. Perhaps it is not necessary, so removed.

AttilaVamos commented on dali/ft/daftformat.cpp:
```diff
@@ -928,6 +928,21 @@ size32_t CCsvPartitioner::getSplitRecordSize(const byte * start, unsigned maxToR
         cur += matchLen;
     }
 
+    numOfProcessedBytes += (unsigned)(end - start);
```
Changed

AttilaVamos commented on testing/regress/ecl/spray_header_test.ecl:
```diff
@@ -0,0 +1,140 @@
+
+//noRoxie
+//noThorLCR
+//noThor
```
Removed

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: a9b8b60bbaca7ef0e641087afc43a2d070b86d23
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.271 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |726 |726 |0 |
|test (thor) |619 |619 |0 |
|test (roxie) |747 |747 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@jakesmith are you happy now?

AttilaVamos commented:
Should I squash before merge?

richardkchapman commented:
Yes

AttilaVamos committed:
HPCC-17506 SPRAY VARIABLE without at least one TERMINATOR fails

Add code to handle the "there is only one row/record and it is
terminated by EOF instead of TERMINATOR and the record structure
definition generation doesn't finish yet" case in the
CCsvPartitioner::getSplitRecordSize() method.

Create Regression Suite test case to test spray a file contains only one
record/row with and without terminator at the end situation.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 1c14aad82bcf19f626fa2b738ccf156804b76b2d
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.955 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |726 |726 |0 |
|test (thor) |619 |619 |0 |
|test (roxie) |747 |747 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented:
@richardkchapman - looks good/ready to merge.
